KAFKA-4213: First set of system tests for replication throttling, KIP-73.

apurvam opened the pull request:
In this patch, we test `kafka-reassign-partitions` when throttling is active. 

This patch also fixes the following:
1. KafkaService.verify_reassign_partitions did not check whether
   partition reassignment actually completed successfully (KAFKA-4204).
   This patch works around those shortcomings so that we get the right
   signal from this method.
2. ProduceConsumeValidateTest.annotate_missing_messages would call
   `pop' on the list of missing messages, causing downstream methods to get
   incomplete data. We fix that in this patch as well.

apurvam committed:
Initial commit of system tests for replication throttling, KIP-73.

This patch also fixes the following:

  1. KafkaService.verify_reassign_partitions did not check whether
partition reassignment actually completed successfully (KAFKA-4204).
This patch works around those shortcomings so that we get the right
signal from this method.

  2. ProduceConsumeValidateTest.annotate_missing_messages would call
`pop' on the list of missing messages, causing downstream methods to get
incomplete data. We fix that in this patch as well.

ijuma commented:
@granders, would have cycles to take a look at this as well?

granders commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,173 @@
+
+
+class ThrottlingTest(ProduceConsumeValidateTest):
+    """
```
Stale docstring (copied from reassign_partitions_test)?
Perhaps unecessary, since you have a detailed docstring in the test method

apurvam replied:
Fixed.

granders commented on tests/kafkatest/services/kafka/kafka.py:
```diff
@@ -138,6 +138,45 @@ def start_minikdc(self, add_principals=""):
         else:
             self.minikdc = None
 
+    def start_some(self, nodes_to_start, create_topics=True, add_principals=""):
```
It doesn't look like this new method is used, can it be deleted?

apurvam replied:
Yes, I think it can be deleted now. It was added because it would be needed for the second item in [CPKAFKA-393](https://confluentinc.atlassian.net/browse/CPKAFKA-393).

But we have since decided that the second item can be rolled into this PR.. @benstopford will submit a patch for that. In which case we don't need this code. Will delete it.

granders commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,173 @@
+        self.logger.debug("Transfer took %d second. Estimated time : %ds",
+                          time_taken,
+                          estimated_throttled_time)
+        assert time_taken >= estimated_throttled_time, \
```
Is it also worth asserting that the relative error (`abs(time_taken- estimated_throttled_time) / time_taken` is within some threshold?

apurvam replied:
@benstopford would be better positioned to answer that. IS the throttling algorithm supposed to be within some range of error?

Based on our conversations, he did not recommend setting an upper bound.

benstopford replied:
We don't mind if it takes longer. Technically it could be a bit shorter though, so for stablity I'd probably suggest something like:  

`time_taken > estimated_throttled_time * 0.9`

apurvam replied:
Actually, the estimate is for the size of the seed data, which should be the bulk of the data. However, we continuously produce messages during reassignment (for a 5 minute run, around 300k small messages are produced and consumed). So the actual time to completion will be a little longer than the estimate, not shorter.

granders commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -28,16 +30,32 @@ class ProduceConsumeValidateTest(Test):
         super(ProduceConsumeValidateTest, self).__init__(test_context=test_context)
+        # How long to wait for the producer to declare itself healthy? This can
+        # be overidden by inheriting classes.
+        self.producer_start_timeout_sec = 60
```
I see this went from 20sec -> 60sec, was it necessary to increase this?

apurvam replied:
Sorry, that is leftover from when I was trying to debug test failures. It doesn't need to be 60 anymore. Will revert to the original 20.

granders commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,173 @@
+
+        self.kafka.start()
+        bulk_producer.run()
+        self.run_produce_consume_validate(core_test_action=
```
Does this test class need to inherit from `ProduceConsumeValidateTest`?
Could you simply bulk-load the data first, and then run `self.reassign_partitions`?

apurvam replied:
Yes, we do need to inherit from `ProduceConsumeValidateTest`.

Part of the test is to make sure that we can produce and consume correctly during throttled reassignment, just like we check for unthrottled reassignment.

granders commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -28,16 +30,32 @@ class ProduceConsumeValidateTest(Test):
+        # timeout avoids is that the consumer is still starting after the
+        # producer begins producing messages, in which case we will miss the
+        # initial set of messages and get spurious test failures.
+        self.delay_between_consumer_and_producer_start_sec = 0
```
You're effectively adding a parametrizable `time.sleep` here, and we want to avoid calls to `sleep`.

It's generally better practice to do a `wait_until` and specify the desired state you want to reach before proceeding.

My understanding is that purpose of this is to guarantee that consumer process is "fully alive and ready to consume" (and this probably doesn't require 10 seconds) before continuing on to start the producer. Can you guarantee this with a `wait_until` with a heuristic check on the state of the consumer process rather than a `time.sleep` call?

apurvam replied:
I will look into this.

apurvam replied:
Switched to using `wait_until`, checking `self.consumer.alive()`.

ijuma replied:
Do we need to rename the variable to include `timeout` there somewhere given the new semantics?

granders commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -80,14 +98,16 @@ def run_produce_consume_validate(self, core_test_action=None, *args):
         else:
             for i in range(20):
-                msg += str(missing.pop()) + ", "
+                msg += str(missing_list[i]) + ", "
```
Could also do
`ms += ", ".join(missing_list[:20])`

apurvam replied:
Done.

apurvam replied:
Actually, the right way to do the join is `", ".join(str(m) for m in missing_list[:20])`, since `missing_list` is a list of ints, and `join` expects strings. Updated the PR with this fix.

granders commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,173 @@
+        return super(ThrottlingTest, self).min_cluster_size() +\
+            self.num_producers + self.num_consumers
+
+    def clean_bounce_some_brokers(self):
```
Looks like this isn't used at the moment - were you planning on toggling `bounce_brokers` at some point?

apurvam replied:
Yes, we will enable bounce_brokers once [KAFKA-4214](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/559/) is fixed. Otherwise, system tests will fail due to that bug and because `KafkaService.verify_reassign_partitions` is fixed in this PR.

apurvam replied:
Broker bounce is enabled in the latest patch.

granders reviewed (changes requested):
Thanks for the tests! I haven't tried running yet, but made a first pass inspection.

Looks good overall but I have several questions and minor suggested tweaks here and there.

benstopford commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,173 @@
+                                  topics={
+                                      self.topic: {
+                                          "partitions": self.num_partitions,
+                                          "replication-factor": 1,
```
It would be best if this test used a replication factor of 2. With a replication factor of 1 we will have no regular replication traffic occurring when the producer writes messages. It would be good to have both throttled replication and non throttled replication happening at the same time.

apurvam replied:
have set replication-factor to 2 now.

benstopford reviewed (commented):
Took a quick pass over this. One comment.

apurvam committed:
Addressed PR comments

apurvam committed:
set replication factor to 2, increase num brokers to 6, decrease partitinos 3, so that we get a predicatble amount of data movement

apurvam committed:
Addressed remaining PR items:

  1. Set replication factor to 2, with related changes to estimated time
     to completion, etc.
  2. Enable bounce of brokers, since the underlying reassignment bug is
     now fixed.

apurvam commented:
I have addressed all outstanding items. Please re-review, thanks!

ijuma commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,171 @@
+
+    Since the correctness is time dependent, this test also simplifies the
+    cluster topology. We have 4 brokers, 1 topic with 4 partitions, and a
+    replication-factor of 1. The reassignment moves every partition. Hence
```
This should be replication-factor of 2 now right?

ijuma replied:
Actually, it seems more than one figure needs updating. Makes me wonder if we should actually remove them altogether and let people read the code (which can't go stale) instead.

apurvam replied:
Agreed. Re wrote the comment to convey the intent, but without the specifics, as they probably will change.

apurvam committed:
Enable the bounce brokers test for reassign partitions as well, since the underlying bug is now fixed

ijuma commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,171 @@
+            num_moves = max(len(new_replicas - old_replicas), num_moves)
+            partition_info["partitions"][i]["partition"] = new_part
+        self.logger.debug("Jumbled partitions: " + str(partition_info))
+        # send reassign partitions command
```
Nitpick: we probably want to have this comment closer to the next command than the previous one. But maybe we should remove it altogether as it doesn't seem to add much over the `execute_reassign_partitions` command.

ijuma commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,171 @@
+        estimated_throttled_time = math.ceil(float(size_per_broker) /
+                                             self.throttle)
+        self.logger.debug("Waiting %ds for the reassignment to complete",
+                          estimated_throttled_time * 2)
```
Probably worth extracting this variable. It seems like it would be easy to update the one in `wait_until` (maybe because we have transient failures) while forgetting to update this one (causing confusing logs).

apurvam replied:
Done.

ijuma commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,171 @@
+            old_replicas = set(partition_info["partitions"][i]["replicas"])
+            new_part = (i+1) % self.num_partitions
+            new_replicas = set(partition_info["partitions"][new_part]["replicas"])
+            num_moves = max(len(new_replicas - old_replicas), num_moves)
```
Would it be clearer to name this `max_num_moves` or something like that?

apurvam replied:
Done.

ijuma reviewed (commented):
Thanks for the PR, left a few minor comments. Would be good for @benstopford to take a second look before we merge it.

apurvam committed:
Fixed comments and variable names, as per Ismael's suggestions.

apurvam committed:
Renamed the `delay_between_producer_and_consumer_start_sec` to
`consumer_init_timeout_sec`. Also introduced
`consumer_start_timeout_sec` to be consistent with the existent
`producer_start_timeout_sec`.

apurvam commented:
> Do we need to rename the variable to include timeout there somewhere given the new semantics?

Somehow I can't reply to the above inline. Here is my response: I renamed it to `consumer_init_timeout_sec`, and also introduce `consumer_start_timeout_sec` to be consistent with the semantics of `producer_start_timeout_sec`. What do you think?

apurvam commented:
Kickstarted the Jenkins system test [here](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/560/).

apurvam commented:
Result of Jenkins branch builder for this run [is here](http://confluent-kafka-branch-builder-system-test-results.s3.amazonaws.com/2016-09-28--001.1475040128--apurvam--throttling-tests--1f55670/report.html).

Only one test failure, that of `TestMirrorMakerService`. Looking at the stack trace, it seems unrelated: it was terminated by a `KeyboardInterrupt` somehow while spinning up zookeeper.

granders commented:
@apurvam FYI this is the job that I cancelled last night with `SIGINT`. The final error shows up as `KeyboardInterrupt` because that is how the python interpreter translates `SIGINT`

apurvam commented:
@granders yep. I just looked at the results, and only 34 tests out of 200+ ran. I kicked of another run a little while back, but it's waiting in the queue.

granders commented:
@apurvam Your branch builder job is now running [here](https://jenkins.confluent.io/job/system-test-kafka-branch-builder/562/console)

ijuma commented:
@apurvam there were some system test failures like the following:

``` text
test_id:    2016-09-28--001.kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=True
status:     FAIL
run time:   2 minutes 12.935 seconds


    {'ProducerPerformanceService-0-140201794908624-worker-1': Exception('No output from ProducerPerformance',)}
Traceback (most recent call last):
```

There are also unrelated failures that were fixed by https://github.com/apache/kafka/commit/a6f3cf56b3ce5c97e89f4b80c34d54ea608950ee

ijuma commented:
I rebased Apurva's branch to include Jason's fix and started a new system tests build (while the US is still sleeping):

https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/103

apurvam committed:
Two fixes:

  1. The ProduceConsumeValidateTest.consumer_init_timeout_sec variable
was not updated in ThrottlingTest. Hence there we did not wait between
starting the consumer and the producer, causing the consumer to miss the
initial messages.

  2. Fix ProduceConsumeValidateTest.annotate_missing_messages: the join
statement was trying to join integers, which is an error. Convert the
integer to a string before applying the join.

apurvam commented:
The reason for the system test failure on `ThrottlingTest` is the following:

```
Exception in thread "main" org.apache.kafka.common.KafkaException: Failed to construct kafka producer
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:339)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:189)
        at org.apache.kafka.tools.ProducerPerformance.main(ProducerPerformance.java:68)
Caused by: org.apache.kafka.common.KafkaException: org.apache.kafka.common.KafkaException: java.io.FileNotFoundException: /mnt/security/test.keystore.jks (No such file or directory)
        at org.apache.kafka.common.network.SslChannelBuilder.configure(SslChannelBuilder.java:44)
        at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:70)
        at org.apache.kafka.clients.ClientUtils.createChannelBuilder(ClientUtils.java:85)
        at org.apache.kafka.clients.producer.KafkaProducer.<init>(KafkaProducer.java:303)
        ... 2 more
Caused by: org.apache.kafka.common.KafkaException: java.io.FileNotFoundException: /mnt/security/test.keystore.jks (No such file or directory)
        at org.apache.kafka.common.security.ssl.SslFactory.configure(SslFactory.java:121)
        at org.apache.kafka.common.network.SslChannelBuilder.configure(SslChannelBuilder.java:41)
        ... 5 more
Caused by: java.io.FileNotFoundException: /mnt/security/test.keystore.jks (No such file or directory)
        at java.io.FileInputStream.open(Native Method)
        at java.io.FileInputStream.<init>(FileInputStream.java:146)
        at java.io.FileInputStream.<init>(FileInputStream.java:101)
        at org.apache.kafka.common.security.ssl.SslFactory$SecurityStore.load(SslFactory.java:216)
        at org.apache.kafka.common.security.ssl.SslFactory$SecurityStore.access$000(SslFactory.java:201)
        at org.apache.kafka.common.security.ssl.SslFactory.createSSLContext(SslFactory.java:137)
        at org.apache.kafka.common.security.ssl.SslFactory.configure(SslFactory.java:119)
        ... 6 more
```

An SSL error even though the tests don't use ssl. This is not reproducible on a non-jenkins system test setup : the tests pass fine.

apurvam commented:
Thanks to @ijuma's efforts, I was able to successfully run this on Jenkins: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/576/

ijuma commented:
@apurvam, the system tests passed. :) Unfortunately, the branch has some conflicts. Can you please resolve them and then just run your tests in the branch builder? I also left a couple of minor comments that I missed last time.

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -32,35 +32,38 @@ def __init__(self, test_context):
         self.consumer.start()
-        if (self.delay_between_consumer_and_producer_start_sec > 0):
+        if (self.consumer_init_timeout_sec > 0):
             self.logger.debug("Sleeping %ds between producer and consumer start",
```
This log message needs to be updated right?

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -32,35 +32,38 @@ def __init__(self, test_context):
             wait_until(lambda: self.consumer.alive(self.consumer.nodes[0]) is True,
-                       timeout_sec=self.delay_between_consumer_and_producer_start_sec,
+                       timeout_sec=self.consumer_init_timeout_sec,
                        err_msg="Consumer process took more than %d s to start" %\
```
Maybe we should say `Consumer process took more than %d s to become alive`? Because the other timeout has the word `start` in the name, it would be good to avoid using it here as it might be confusing.

apurvam replied:
Updated it to `took more than %d s to fork`, which is what is actually happening.

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -32,35 +32,38 @@ def __init__(self, test_context):
         # producer begins producing messages, in which case we will miss the
         # initial set of messages and get spurious test failures.
-        self.delay_between_consumer_and_producer_start_sec = 0
+        self.consumer_init_timeout_sec = 0
```
Would this be better as `consumer_alive_timeout_sec`? Also the documentation above needs to be updated.

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -32,35 +32,38 @@ def __init__(self, test_context):
         wait_until(lambda: len(self.consumer.messages_consumed[1]) > 0,
-                   timeout_sec=60,
+                   timeout_sec=self.consumer_start_timeout_sec,
                    err_msg="Consumer failed to consumer messages for %d s." % 60)
```
`consumer messages` should be `consume messages`

apurvam replied:
Fixed.

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -28,18 +28,43 @@ class ProduceConsumeValidateTest(Test):
+        # timeout avoids is that the consumer is still starting after the
+        # producer begins producing messages, in which case we will miss the
+        # initial set of messages and get spurious test failures.
+        self.consumer_init_timeout_sec = 0
```
Would this be better as consumer_alive_timeout_sec? Also the comment above needs to be updated.

apurvam replied:
Not sure about this. The choice of what is more accurate between `init` and `alive` is subjective. As such, doesn't make sense to me to change the name either way.

I updated the comment.

ijuma replied:
I suggested `alive` to be consistent with the method being called on the consumer. The main reason is that `init` and `start` are a bit too similar and it's a bit difficult to distinguish between the two. Anyway, ok to leave as is if you feel strongly about it.

apurvam replied:
I actually don't feel strongly that one or the other makes things any clearer. As such, I'd prefer to leave the name as is.

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -28,18 +28,43 @@ class ProduceConsumeValidateTest(Test):
-             err_msg="Consumer failed to start in a reasonable amount of time.")
+        if (self.consumer_init_timeout_sec > 0):
+            self.logger.debug("Sleeping %ds between producer and consumer start",
+                              self.consumer_init_timeout_sec)
```
This log message needs to be updated right?

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -28,18 +28,43 @@ class ProduceConsumeValidateTest(Test):
+                              self.consumer_init_timeout_sec)
+            wait_until(lambda: self.consumer.alive(self.consumer.nodes[0]) is True,
+                       timeout_sec=self.consumer_init_timeout_sec,
+                       err_msg="Consumer process took more than %d s to start" %\
```
This log message needs to be updated right?

apurvam replied:
Done.

ijuma commented on tests/kafkatest/tests/produce_consume_validate.py:
```diff
@@ -28,18 +28,43 @@ class ProduceConsumeValidateTest(Test):
+                   err_msg="Producer failed to start in a reasonable amount of time.")
+        wait_until(lambda: len(self.consumer.messages_consumed[1]) > 0,
+                   timeout_sec=self.consumer_start_timeout_sec,
+                   err_msg="Consumer failed to consumer messages for %d s." % 60)
```
This log message needs to be updated right?

apurvam replied:
done.

ijuma commented on tests/kafkatest/tests/core/throttling_test.py:
```diff
@@ -0,0 +1,174 @@
+        # seeding the cluster with big data (to test throttling), we need to
+        # Start the consumer from the end of the stream. further, we need to
+        # ensure that the consumer is fully started before the producer starts
+        # so that we don't miss any messages. This delay ensures the sufficient
```
"delay" -> "timeout"?

apurvam replied:
done.

apurvam committed:
Addressed PR comments:

 1. Fixed outdated code comments.
 2. Fixed outdated log messages.

apurvam commented:
The system tests on jenkins with the latest set of patches + rebase passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/578/
http://confluent-kafka-branch-builder-system-test-results.s3.amazonaws.com/2016-09-30--001.1475253141--apurvam--throttling-tests--8fb2eb9/report.html

ijuma commented:
Thanks for the PR, LGTM. Merged to trunk and 0.10.1 branches.
